lib/Fmt_ast.ml

```diff
@@ -2893,6 +2893,21 @@ and fmt_expression c ?(box = true) ?(pro = noop) ?eol ?parens
   | Pexp_indexop_access x ->
       pro $ fmt_indexop_access c ctx ~fmt_atrs ~has_attr ~parens x
   | Pexp_hole -> pro $ hvbox 0 (fmt_hole () $ fmt_atrs)
+  | Pexp_beginend
+      ( { pexp_desc= Pexp_function (args, typ, body)
+        ; pexp_attributes= function_attrs
+        ; _ } as f ) ->
+      let wrap_intro x =
+        hvbox 0
+          ( hvbox 0 (str "begin" $ fmt_extension_suffix c ext $ fmt_atrs)
+          $ break 1 2 $ x )
+        $ break 1000 0
+      in
+      hvbox 2
+        (fmt_function ~wrap_intro ~box ~ctx:(Exp f) ~ctx0 ~label:Nolabel
```

Comment:
This makes the formatting of `Pexp_function` too complex by duplicating codes with subtle changes. `fmt_function` is extremely complex for exactly this reason.
This won't scale as more `begin..end` special cases are added (for example https://github.com/ocaml-ppx/ocamlformat/pull/2666) and might create a mess.
This could be solved by passing the `~pro` attribute. I have tried this in this branch but there are still some unwanted diffs: https://github.com/ocaml-ppx/ocamlformat/compare/main...Julow:ocamlformat:2664_passing_pro?expand=1
Can you have a look ?

Reply:
This is adressed.